File: samples/graph-indirect-membership-change-notification/python/helper/graph_helper.py
import os
import requests
from helper.auth import AuthHelper


class GraphHelper:
    @staticmethod
    def get_headers(token):
        return {
            "Accept": "application/json",
            "Content-Type": "application/json",
            "Authorization": f"Bearer {token}"
        }

    @staticmethod
    def delete_subscription(subscription_id):
        token = AuthHelper.get_access_token()
        url = f"https://graph.microsoft.com/v1.0/subscriptions/{subscription_id}"
        headers = GraphHelper.get_headers(token)
        response = requests.delete(url, headers=headers)
        response.raise_for_status()

    @staticmethod
    def create_subscription(teams_id, page_id):
        token = AuthHelper.get_access_token()
        notification_url = os.getenv("notificationUrl")
        resource = ""
        change_type = ""

        if page_id == "1":
            resource = f"/teams/{teams_id}/channels/getAllMembers"
            change_type = "created,deleted,updated"

        headers = GraphHelper.get_headers(token)

        try:
            resp = requests.get("https://graph.microsoft.com/beta/subscriptions", headers=headers)
            existing_subscriptions = resp.json().get("value", [])
        except requests.RequestException:
            return None

        existing_subscription = next(
            (sub for sub in existing_subscriptions if sub["resource"] == resource), None
        )

        if existing_subscription and existing_subscription["notificationUrl"] != notification_url:
            print(f"CreateNewSubscription-ExistingSubscriptionFound: {resource}")
            GraphHelper.delete_subscription(existing_subscription["id"])
            existing_subscription = None

        if not existing_subscription:
            subscription_payload = {
                "changeType": change_type,
                "notificationUrl": notification_url,
                "lifecycleNotificationUrl": notification_url,
                "resource": resource,
                "includeResourceData": True,
                "encryptionCertificate": os.getenv("Base64EncodedCertificate"),
                "encryptionCertificateId": "meeting-notification",
                "expirationDateTime": GraphHelper.get_expiration_time(hours=10),
                "clientState": "clientState"
            }

            try:
                resp = requests.post(
                    "https://graph.microsoft.com/v1.0/subscriptions",
                    headers=headers,
                    json=subscription_payload
                )
                existing_subscription = resp.json()
            except requests.RequestException as e:
                print(f"Error--{e}")
                return None

        return existing_subscription

    @staticmethod
    def create_shared_with_teams_subscription(teams_id, channel_id):
        token = AuthHelper.get_access_token()
        resource = f"/teams/{teams_id}/channels/{channel_id}/sharedWithTeams"
        change_type = "created,deleted,updated"
        notification_url = os.getenv("notificationUrl")
        headers = GraphHelper.get_headers(token)

        try:
            resp = requests.get("https://graph.microsoft.com/beta/subscriptions", headers=headers)
            existing_subscriptions = resp.json().get("value", [])
        except requests.RequestException:
            return None

        existing_subscription = next(
            (sub for sub in existing_subscriptions if sub["resource"] == resource), None
        )

        if existing_subscription and existing_subscription["notificationUrl"] != notification_url:
            GraphHelper.delete_subscription(existing_subscription["id"])
            existing_subscription = None

        if not existing_subscription:
            subscription_payload = {
                "changeType": change_type,
                "notificationUrl": notification_url,
                "lifecycleNotificationUrl": notification_url,
                "resource": resource,
                "includeResourceData": True,
                "encryptionCertificate": os.getenv("Base64EncodedCertificate"),
                "encryptionCertificateId": "meeting-notification",
                "expirationDateTime": GraphHelper.get_expiration_time(days=50) + 'Z',
                "clientState": "clientState"
            }

            try:
                resp = requests.post(
                    "https://graph.microsoft.com/beta/subscriptions",
                    headers=headers,
                    json=subscription_payload
                )
                existing_subscription = resp.json()
            except requests.RequestException as e:
                print(f"Error--{e}")
                return None

        return existing_subscription

    @staticmethod
    def check_user_channel_access(team_id, channel_id, user_id, tenant_id):
        token = AuthHelper.get_access_token()
        url = (
            f"https://graph.microsoft.com/v1.0/teams/{team_id}/channels/{channel_id}/"
            f"microsoft.graph.doesUserHaveAccess(userId='{user_id}', tenantId='{tenant_id}')"
        )

        try:
            response = requests.get(url, headers=GraphHelper.get_headers(token))
            return response.json().get("value", False)
        except requests.RequestException as e:
            print(f"Error checking user channel access: {e}")
            return False

    @staticmethod
    def get_expiration_time(hours=0, days=0):
        from datetime import datetime, timedelta
        return (datetime.utcnow() + timedelta(hours=hours, days=days)).isoformat() + "Z"

File: samples/graph-indirect-membership-change-notification/nodejs/helper/graph-helper.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const auth = require("../auth");
const axios = require('axios');

class GraphHelper {
    /**
     * Delete existing subscription.
     * @param {string} subscriptionId Id of subscription to be deleted.
     */
    static async deleteSubscription(subscriptionId) {
        var applicationToken = await auth.getAccessToken();

        await axios.delete(`https://graph.microsoft.com/v1.0/subscriptions/${subscriptionId}`, {
            headers: {
                "accept": "application/json",
                "contentType": 'application/json',
                "authorization": "bearer " + applicationToken
            }
        });
    }

    /**
    * PageId =1 for Channel Subscription
    * PageId =2 for Team Subscription.
    * @param {pageId}. Team and channel subscription.
    */
    static async createSubscription(teamsId, pageId) {
        debugger;
        let applicationToken = await auth.getAccessToken();
        let resource = "";
        let changeType = "";
        let notificationUrl = process.env.notificationUrl;

        if (pageId === "1") {
            resource = `/teams/${teamsId}/channels/getAllMembers`;
            changeType = "created,deleted,updated";
        }
        else {
            resource = `/teams/${teamsId}`;
            changeType = "deleted,updated"
        }

        let existingSubscriptions = null;

        try {
            var apiResponse = await axios.get(`https://graph.microsoft.com/beta/subscriptions`, {
                headers: {
                    "accept": "application/json",
                    "contentType": 'application/json',
                    "authorization": "bearer " + applicationToken
                }
            });

            existingSubscriptions = apiResponse.data.value;
        }
        catch (ex) {
            return null;
        }

        var existingSubscription = existingSubscriptions.find(subscription => subscription.resource === resource);

        if (existingSubscription != null && existingSubscription.notificationUrl != notificationUrl) {
            console.log(`CreateNewSubscription-ExistingSubscriptionFound: ${resource}`);
            await this.deleteSubscription(existingSubscription.id);
            existingSubscription = null;
        }

        try {            
            
            if (existingSubscription == null) {

                let subscriptionCreationInformation = {
                    changeType: changeType,
                    notificationUrl: notificationUrl,
                    lifecycleNotificationUrl: notificationUrl, // Adding lifecycle notification URL
                    resource: resource,
                    includeResourceData: true,
                    encryptionCertificate: process.env.Base64EncodedCertificate,
                    encryptionCertificateId: "meeting-notification",
                    expirationDateTime: new Date(Date.now() + 36000000).toISOString(),
                    clientState: "clientState"
                };

                var response = await axios.post(`https://graph.microsoft.com/beta/subscriptions`,subscriptionCreationInformation, {
                    headers: {
                        "accept": "application/json",
                        "contentType": 'application/json',
                        "authorization": "bearer " + applicationToken
                    }
                });

                existingSubscription = response.data;
            }
        }
        catch (e) {
            console.log("Error--" + e);
        }

        return existingSubscription;
    }

    /**
    * Create a subscription for sharedWithTeams on a specific channel.
    * @param {string} teamsId Team ID
    * @param {string} channelId Channel ID
    */
    static async createSharedWithTeamsSubscription(teamsId, channelId) {
        let applicationToken = await auth.getAccessToken();
        let resource = `/teams/${teamsId}/channels/${channelId}/sharedWithTeams`;
        let changeType = "created,deleted,updated";
        let notificationUrl = process.env.notificationUrl;
        let existingSubscriptions = null;

        try {
            var apiResponse = await axios.get(`https://graph.microsoft.com/beta/subscriptions`, {
                headers: {
                    "accept": "application/json",
                    "contentType": 'application/json',
                    "authorization": "bearer " + applicationToken
                }
            });
            existingSubscriptions = apiResponse.data.value;
        } catch (ex) {
            return null;
        }

        var existingSubscription = existingSubscriptions.find(subscription => subscription.resource === resource);

        if (existingSubscription != null && existingSubscription.notificationUrl != notificationUrl) {
            await this.deleteSubscription(existingSubscription.id);
            existingSubscription = null;
        }

        try {            if (existingSubscription == null) {
                let subscriptionCreationInformation = {
                    changeType: changeType,
                    notificationUrl: notificationUrl,
                    lifecycleNotificationUrl: notificationUrl, // Adding lifecycle notification URL
                    resource: resource,
                    includeResourceData: true,
                    encryptionCertificate: process.env.Base64EncodedCertificate,
                    encryptionCertificateId: "meeting-notification",
                    expirationDateTime: new Date(Date.now() + 4320000000).toISOString(), // Set to 50 days
                    clientState: "clientState"
                };
                var response = await axios.post(`https://graph.microsoft.com/beta/subscriptions`,subscriptionCreationInformation, {
                    headers: {
                        "accept": "application/json",
                        "contentType": 'application/json',
                        "authorization": "bearer " + applicationToken
                    }
                });
                existingSubscription = response.data;
            }
        } catch (e) {
            console.log("Error--" + e);
        }
        return existingSubscription;
    }    
    
    /**
     * Check if a user has access to a specific channel
     * @param {string} teamId Team ID
     * @param {string} channelId Channel ID
     * @param {string} userId User ID
     * @param {string} tenantId Tenant ID
     * @returns {Promise<boolean>} True if the user has access, false otherwise
     */
    static async checkUserChannelAccess(teamId, channelId, userId, tenantId) {
        try {
            const applicationToken = await auth.getAccessToken();
            const response = await axios.get(
                `https://graph.microsoft.com/v1.0/teams/${teamId}/channels/${channelId}/microsoft.graph.doesUserHaveAccess(userId='${userId}', tenantId='${tenantId}')`,
                {
                    headers: {
                        "accept": "application/json",
                        "authorization": "bearer " + applicationToken
                    }
                }
            );
            return response.data.value;
        } catch (error) {
            console.log("Error checking user channel access:", error);
            return false;
        }
    }
}

exports.GraphHelper = GraphHelper;